### PR TITLE
Remove Linq and boxed Enumerator allocations from conneg

### PR DESCRIPTION
**Pull request by rynowak**

### CODE REVIEW

**Comment by rynowak** [comment]
For #989

File: src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs [hunk 1]
```diff
@@ -273,5 +271,63 @@ private Encoding MatchAcceptCharacterEncoding(IList<StringWithQualityHeaderValue
...
+            for (var i = 0; i < values.Count; i++)
+            {
+                var value = values[i];
+                if (value.Quality == HeaderQuality.NoMatch)
+                {
+                    // Exclude this one
+                }
+                else
+                {
+                    // Go read the documentation, I'm tired of explaining it.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs]: A fine comment for your co-workers but should not be in the code. Either leave this out or provide the MSDN link of interest.

File: src/Microsoft.AspNet.Mvc.Core/Infrastructure/ObjectResultExecutor.cs [hunk 2]
```diff
@@ -180,6 +179,8 @@ public virtual Task ExecuteAsync(ActionContext context, ObjectResult result)
                 throw new ArgumentNullException(nameof(formatters));
             }
 
+            var request = formatterContext.HttpContext.Request;
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/ObjectResultExecutor.cs]: Move down before 1st use.

File: src/Microsoft.AspNet.Mvc.Core/Infrastructure/ObjectResultExecutor.cs [hunk 3]
```diff
@@ -397,60 +412,114 @@ public virtual Task ExecuteAsync(ActionContext context, ObjectResult result)
...
+                    var message = Resources.FormatObjectResult_MatchAllContentType(
+                        contentType,
+                        nameof(ObjectResult.ContentTypes));
+                    throw new InvalidOperationException(message);
+                }
             }
         }
 
-        // This can't be cached, because 
+        // This can't be cached, because the BindingContext is per-request.
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/ObjectResultExecutor.cs]: :clap:

File: src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs [hunk 4]
```diff
@@ -251,21 +250,20 @@ private Encoding MatchAcceptCharacterEncoding(IList<StringWithQualityHeaderValue
         {
             if (acceptCharsetHeaders != null && acceptCharsetHeaders.Count > 0)
             {
-                var sortedAcceptCharsetHeaders = acceptCharsetHeaders
-                    .Where(acceptCharset => acceptCharset.Quality != HeaderQuality.NoMatch)
-                    .OrderByDescending(m => m, StringWithQualityHeaderValueComparer.QualityComparer);
-
-                foreach (var acceptCharset in sortedAcceptCharsetHeaders)
+                var sorted = Sort(acceptCharsetHeaders);
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs]: Suggest `acceptValues` is better here as well. Liked that name in the next file.

File: src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs [hunk 5]
```diff
@@ -273,5 +271,63 @@ private Encoding MatchAcceptCharacterEncoding(IList<StringWithQualityHeaderValue
...
+                        sorted.Insert(position + 1, value);
+                    }
+                    else
+                    {
+                        sorted.Insert(~position, value);
+                    }
+                }
+            }
+
+            // We want a descending sort, but BinarySearch does ascending
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs]: Is the reversal costly enough to create an inverted quality comparer?

File: src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs [hunk 6]
```diff
@@ -273,5 +271,63 @@ private Encoding MatchAcceptCharacterEncoding(IList<StringWithQualityHeaderValue
 
             return null;
         }
+
+        // There's no allocation-free way to sort an IList and we may have to filter anyway,
+        // so we're going to have to live with the copy + insertion sort.
+        private IList<StringWithQualityHeaderValue> Sort(IList<StringWithQualityHeaderValue> values)
```
Comment (dougbu) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs]: Worth creating a `IList<T> Sort<T>(IList<T>, IComparer<T>)` method and sharing that?

**Comment by dougbu** [comment]
:shipit: w/ a few minor fixes. LMK if you make larger changes in answer to my questions and I'll :trollface: back through.

File: src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs (hunk 6 above)
Comment (pranavkm) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs]: `static`?

Comment (NeelBhatt) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs]: something like 

```
public static void Sort<T>(
	IList<T> data,
	IComparer<T> comparer
)
```

File: src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs (hunk 1 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs]: :trollface:

File: src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs (hunk 5 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs]: If the data show that then we will.

File: src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs (hunk 6 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs]: :-1: - Ideally we don't have to have/maintain that kind of code.

File: src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs (hunk 4 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Formatters/OutputFormatter.cs]: sure

File: src/Microsoft.AspNet.Mvc.Core/Infrastructure/ObjectResultExecutor.cs (hunk 2 above)
Comment (rynowak) [inline src/Microsoft.AspNet.Mvc.Core/Infrastructure/ObjectResultExecutor.cs]: sure

**Pull request merged**